bip-p2qrh.mediawiki

```diff
@@ -0,0 +1,587 @@
+
+==== Hash Computation ====
+
+hash = HASH256(HASH256(pubkey1) || HASH256(pubkey2) || ... || HASH256(pubkeyN))
+
+==== Sighash Calculation ====
+
+The sighash for P2QRH outputs follows the same procedure as defined in BIP-0143 for SegWit transactions:
+
+* '''Hash Prevouts:''' Computed over the previous outputs being spent.
+* '''Hash Sequence:''' Computed over the sequence fields.
+* '''Hash Outputs:''' Computed over the outputs of the transaction.
+
+The message to be signed includes these hashes, ensuring transaction malleability is prevented.
+
+==== Signature Verification ====
+
+Signature verification is as follows:
+
+1. Extract the <nowiki><hash></nowiki> from the <code>scriptPubKey</code>.
+
+2. For each input:
+
+* Compute <code>hashed_pubkeys</code> by concatenating the <code>HASH256</code> of each provided public key:
+
+  hashed_pubkeys = HASH256(pubkey1) || HASH256(pubkey2) || ... || HASH256(pubkeyN)
+
+* Compute <code>computed_hash</code>:
+
+  computed_hash = HASH256(hashed_pubkeys)
+
+* Compare the resulting hash to <nowiki><hash></nowiki>. If they do not match, the script fails.
+
+3. Verify each signature against the corresponding public key and the sighash.
+
+4. Ensure that the signature algorithm used matches the expected lengths for NIST Level V security and is supported.
+
+==== Attestation Parsing Example ====
+
+Signing for a single input using both FALCON-1024 and secp256k1 Schnorr:
+
+Number of public keys:
+
+  [num_pubkeys]: 0x02
+
+Pubkey 1:
+  [pubkey_length]: 0x0701 (1793 bytes)
+  [pubkey]: public_key_falcon_1024
+
+Pubkey 2:
+  [pubkey_length]: 0x20 (32 bytes)
+  [pubkey]: public_key_secp256k1
+
+Number of signatures:
+
+  [num_signatures]: 0x02
+
+Signature 1:
+  [signature_length]: 0x0500 (1280 bytes)
+  [signature]: signature_falcon_1024
+
+Signature 2:
+  [signature_length]: 0x40 (64 bytes)
+  [signature]: signature_secp256k1
+
+Note: This contrasts with multisig inputs, where the attestation structure repeats for each public key and signature.
+
+=== Compatibility with BIP-141 ===
+
+By adhering to the SegWit transaction structure and versioning, P2QRH outputs are compatible with existing transaction
+processing rules. Nodes that do not recognize SegWit version 3 will treat these outputs as anyone-can-spend but, per
+BIP-141, will not relay or mine such transactions.
+
+=== Usage Considerations ===
+
+==== Transaction Size and Fees ====
+
+Quantum-resistant signatures are significantly larger than traditional signatures, increasing transaction size and the
+fees required. Users and wallet developers should be aware of this and plan accordingly.
+
+For example, for CRYSTALS-Dilithium Level V, a single signature is 4,595 bytes, a substantial increase over current
+ECDSA or Schnorr signatures.
+
+==== Performance Impact ====
+
+Verification of quantum-resistant signatures will be computationally more intensive, and any attestation discount will
```

Comment:
What about using a STARK over all the post-quantum signatures included in a block by the miner to accelerate verification?

Reply:
As we discussed privately, a STARKS-based approach would be out of scope for this BIP, but alternative approaches specified in separate BIPs are certainly welcome.